File: training/llmviz_train/rwkv.py
"""Minimal RWKV-4 reference implementation with activation recording.

Golden source for the TypeScript port: every recorded intermediate is
compared against the TS re-implementation at 1e-4. Clarity beats speed —
the WKV recurrence is a plain Python loop over time, one step per token,
in the numerically-stable max-shifted form used by the RWKV CUDA kernel.

Architecture (two residual sub-blocks per layer, pre-LayerNorm):

    tokens -> embedding
           -> [ x = x + time_mix(ln1(x)) ;  x = x + channel_mix(ln2(x)) ] * n_layer
           -> ln_out -> head (untied) -> logits

Time-mixing (linear attention with per-channel decay):
    token-shift mixes x with the previous position; r,k,v are linear maps of
    the mixed input; the WKV recurrence keeps a numerator/denominator state
    per channel with decay w = -exp(time_decay) and bonus u = time_first;
    out = (sigmoid(r) * wkv) @ output.

Channel-mixing (gated squared-ReLU MLP):
    token-shift; k = relu(xk @ key)^2 (width d_ffn); out = sigmoid(xr @ rec) * (k @ value).
"""

import torch
import torch.nn.functional as F
from torch import nn

from llmviz_train.recording import snapshot


def _token_shift(x: torch.Tensor) -> torch.Tensor:
    """Each position sees the previous position's vector (zeros before t=0)."""
    return F.pad(x, (0, 0, 1, -1))  # shift along time (dim=1) by one, zero-fill


class RWKVTimeMix(nn.Module):
    """WKV linear-attention sub-block (no residual; the model adds it)."""

    def __init__(self, d_model: int):
        super().__init__()
        self.time_decay = nn.Parameter(torch.linspace(-3.0, 3.0, d_model))
        self.time_first = nn.Parameter(torch.zeros(d_model))
        self.time_mix_k = nn.Parameter(torch.full((d_model,), 0.5))
        self.time_mix_v = nn.Parameter(torch.full((d_model,), 0.5))
        self.time_mix_r = nn.Parameter(torch.full((d_model,), 0.5))
        self.key = nn.Linear(d_model, d_model, bias=False)
        self.value = nn.Linear(d_model, d_model, bias=False)
        self.receptance = nn.Linear(d_model, d_model, bias=False)
        self.output = nn.Linear(d_model, d_model, bias=False)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) (already LayerNorm'd) -> (B, T, d_model)."""
        B_, T, C = x.shape

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = snapshot(tensor)

        xs = _token_shift(x)
        xk = x * self.time_mix_k + xs * (1 - self.time_mix_k)
        xv = x * self.time_mix_v + xs * (1 - self.time_mix_v)
        xr = x * self.time_mix_r + xs * (1 - self.time_mix_r)

        r = torch.sigmoid(self.receptance(xr))  # (B, T, C)
        k = self.key(xk)
        v = self.value(xv)
        rec("r", r[0])
        rec("k", k[0])
        rec("v", v[0])

        w = -torch.exp(self.time_decay)  # (C,) per-channel decay, negative
        u = self.time_first  # (C,) bonus for the current token

        # Numerically-stable WKV: keep (aa, bb, pp) state per channel, with pp
        # the running max-exponent so the exponentials never overflow.
        aa = x.new_zeros(B_, C)
        bb = x.new_zeros(B_, C)
        pp = x.new_full((B_, C), -1e38)
        wkv = x.new_zeros(B_, T, C)
        for t in range(T):
            kt = k[:, t]
            vt = v[:, t]
            ww = u + kt
            q = torch.maximum(pp, ww)
            e1 = torch.exp(pp - q)
            e2 = torch.exp(ww - q)
            wkv[:, t] = (e1 * aa + e2 * vt) / (e1 * bb + e2)
            ww = pp + w
            q = torch.maximum(ww, kt)
            e1 = torch.exp(ww - q)
            e2 = torch.exp(kt - q)
            aa = e1 * aa + e2 * vt
            bb = e1 * bb + e2
            pp = q
        rec("wkv", wkv[0])

        out = self.output(r * wkv)
        rec("out", out[0])
        return out


class RWKVChannelMix(nn.Module):
    """Gated squared-ReLU MLP sub-block (no residual; the model adds it)."""

    def __init__(self, d_model: int, d_ffn: int):
        super().__init__()
        self.time_mix_k = nn.Parameter(torch.full((d_model,), 0.5))
        self.time_mix_r = nn.Parameter(torch.full((d_model,), 0.5))
        self.key = nn.Linear(d_model, d_ffn, bias=False)
        self.receptance = nn.Linear(d_model, d_model, bias=False)
        self.value = nn.Linear(d_ffn, d_model, bias=False)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) (already LayerNorm'd) -> (B, T, d_model)."""

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = snapshot(tensor)

        xs = _token_shift(x)
        xk = x * self.time_mix_k + xs * (1 - self.time_mix_k)
        xr = x * self.time_mix_r + xs * (1 - self.time_mix_r)

        k = torch.square(torch.relu(self.key(xk)))  # (B, T, d_ffn)
        rec("k", k[0])
        r = torch.sigmoid(self.receptance(xr))
        out = r * self.value(k)
        rec("out", out[0])
        return out


class RWKV(nn.Module):
    """Minimal RWKV-4 language model over a small token vocabulary."""

    def __init__(self, cfg: dict, vocab_size: int):
        super().__init__()
        n_layer = cfg["n_layer"]
        d_model = cfg["d_model"]
        d_ffn = cfg["ffn_mult"] * d_model

        self.embedding = nn.Embedding(vocab_size, d_model)
        self.ln1 = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.ln2 = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.att = nn.ModuleList(RWKVTimeMix(d_model) for _ in range(n_layer))
        self.ffn = nn.ModuleList(RWKVChannelMix(d_model, d_ffn) for _ in range(n_layer))
        self.ln_out = nn.LayerNorm(d_model)
        self.head = nn.Linear(d_model, vocab_size, bias=False)  # untied

    def forward(
        self, tokens: torch.Tensor, record: bool = False
    ) -> torch.Tensor | tuple[torch.Tensor, dict[str, torch.Tensor]]:
        """tokens: (B, T) long -> logits (B, T, vocab).

        With ``record=True``, also returns a dict of named activations for
        batch item 0 (detached float32 CPU tensors).
        """
        acts: dict[str, torch.Tensor] | None = {} if record else None

        def rec(name: str, tensor: torch.Tensor) -> None:
            if acts is not None:
                acts[name] = snapshot(tensor)

        x = self.embedding(tokens)
        rec("embed.out", x[0])

        for i in range(len(self.att)):
            n1 = self.ln1[i](x)
            rec(f"layer{i}.ln1.out", n1[0])
            x = x + self.att[i](n1, record=acts, prefix=f"layer{i}.att.")
            n2 = self.ln2[i](x)
            rec(f"layer{i}.ln2.out", n2[0])
            x = x + self.ffn[i](n2, record=acts, prefix=f"layer{i}.ffn.")

        x = self.ln_out(x)
        rec("final_norm.out", x[0])

        logits = self.head(x)
        rec("head.logits", logits[0])

        if acts is not None:
            return logits, acts
        return logits

File: src/compute/rwkv.ts
/**
 * TS port of the RWKV-4 forward pass in training/llmviz_train/rwkv.py,
 * operation-for-operation, recording the same activation names. Gated by
 * rwkv.golden.test.ts against PyTorch goldens at 1e-4.
 *
 * Per layer, two residual sub-blocks over a pre-LayerNorm'd input:
 *   time-mix: token-shift → r,k,v → WKV recurrence → out = (σ(r)·wkv)·Wo
 *   channel-mix: token-shift → k = relu(·)² → out = σ(r) · (k·Wv)
 *
 * WKV is the numerically-stable max-shifted recurrence (one step per token),
 * matching the RWKV CUDA kernel and the Python reference loop.
 */

import { T } from "./tensor";
import { add, embedding, layernorm, linear, mul } from "./ops";
import { getTensor, type Manifest } from "./loader";
import type { Recorder } from "./recorder";
import { dimsFromManifest } from "./util";

export interface RwkvDims {
  n_layer: number;
  d_model: number;
  ffn_mult: number;
  vocab_size: number;
  d_ffn: number;
}

const RWKV_DIM_KEYS = ["n_layer", "d_model", "ffn_mult", "vocab_size", "d_ffn"] as const;

/** Validated RwkvDims from a manifest (throws on wrong arch / missing keys). */
export function rwkvDimsFrom(manifest: Manifest): RwkvDims {
  return dimsFromManifest(manifest, "rwkv", RWKV_DIM_KEYS);
}

function sigmoid(x: T): T {
  const out = new Float32Array(x.data.length);
  for (let i = 0; i < out.length; i++) out[i] = 1 / (1 + Math.exp(-x.data[i]));
  return T.from(out, x.shape);
}

/** relu then square (RWKV channel-mix hidden activation). */
function reluSquare(x: T): T {
  const out = new Float32Array(x.data.length);
  for (let i = 0; i < out.length; i++) {
    const v = x.data[i] > 0 ? x.data[i] : 0;
    out[i] = v * v;
  }
  return T.from(out, x.shape);
}

/** Each row (position) sees the previous row; row 0 is zeros. x: (T, C). */
function tokenShift(x: T): T {
  const [steps, c] = x.shape;
  const out = new Float32Array(steps * c); // zero-filled row 0
  for (let t = 1; t < steps; t++) out.set(x.data.subarray((t - 1) * c, t * c), t * c);
  return T.from(out, x.shape);
}

/** Per-channel mix: out[t,c] = x·mix[c] + xs·(1−mix[c]). x,xs: (T,C); mix: (C,). */
function timeMix(x: T, xs: T, mix: T): T {
  const [steps, c] = x.shape;
  const out = new Float32Array(steps * c);
  for (let t = 0; t < steps; t++) {
    for (let j = 0; j < c; j++) {
      const m = mix.data[j];
      const i = t * c + j;
      out[i] = x.data[i] * m + xs.data[i] * (1 - m);
    }
  }
  return T.from(out, x.shape);
}

/**
 * WKV linear attention: per channel, a numerator/denominator state with
 * decay w = −exp(time_decay) and current-token bonus u = time_first, in the
 * max-shifted stable form. k, v: (T, C); time_decay, time_first: (C,).
 */
function wkv(k: T, v: T, timeDecay: T, timeFirst: T): T {
  const [steps, c] = k.shape;
  const out = new Float32Array(steps * c);
  const aa = new Float32Array(c);
  const bb = new Float32Array(c);
  const pp = new Float32Array(c).fill(-1e38);
  const w = new Float32Array(c);
  for (let j = 0; j < c; j++) w[j] = -Math.exp(timeDecay.data[j]);
  const u = timeFirst.data;

  for (let t = 0; t < steps; t++) {
    const base = t * c;
    for (let j = 0; j < c; j++) {
      const kt = k.data[base + j];
      const vt = v.data[base + j];
      // output for the current token (includes the bonus u)
      let ww = u[j] + kt;
      let q = Math.max(pp[j], ww);
      let e1 = Math.exp(pp[j] - q);
      let e2 = Math.exp(ww - q);
      out[base + j] = (e1 * aa[j] + e2 * vt) / (e1 * bb[j] + e2);
      // advance the state by one step of decay
      ww = pp[j] + w[j];
      q = Math.max(ww, kt);
      e1 = Math.exp(ww - q);
      e2 = Math.exp(kt - q);
      aa[j] = e1 * aa[j] + e2 * vt;
      bb[j] = e1 * bb[j] + e2;
      pp[j] = q;
    }
  }
  return T.from(out, k.shape);
}

/** One time-mixing sub-block (no residual). x already LayerNorm'd: (T, d_model). */
function timeMixBlock(
  weights: Map<string, T>,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const p = `layer${layer}.att.`;
  const w = (s: string): T => getTensor(weights, `att.${layer}.${s}`);
  const xs = tokenShift(x);
  const r = sigmoid(linear(timeMix(x, xs, w("time_mix_r")), w("receptance.weight")));
  const k = linear(timeMix(x, xs, w("time_mix_k")), w("key.weight"));
  const v = linear(timeMix(x, xs, w("time_mix_v")), w("value.weight"));
  rec?.record(`${p}r`, r);
  rec?.record(`${p}k`, k);
  rec?.record(`${p}v`, v);
  const wk = wkv(k, v, w("time_decay"), w("time_first"));
  rec?.record(`${p}wkv`, wk);
  const out = linear(mul(r, wk), w("output.weight"));
  rec?.record(`${p}out`, out);
  return out;
}

/** One channel-mixing sub-block (no residual). x already LayerNorm'd: (T, d_model). */
function channelMixBlock(
  weights: Map<string, T>,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const p = `layer${layer}.ffn.`;
  const w = (s: string): T => getTensor(weights, `ffn.${layer}.${s}`);
  const xs = tokenShift(x);
  const k = reluSquare(linear(timeMix(x, xs, w("time_mix_k")), w("key.weight")));
  rec?.record(`${p}k`, k);
  const r = sigmoid(linear(timeMix(x, xs, w("time_mix_r")), w("receptance.weight")));
  const out = mul(r, linear(k, w("value.weight")));
  rec?.record(`${p}out`, out);
  return out;
}

/**
 * Full forward: tokens → logits (T, vocab), recording the same activation
 * names as rwkv.py when `rec` is given.
 */
export function rwkvForward(
  weights: Map<string, T>,
  dims: RwkvDims,
  tokens: number[],
  rec?: Recorder,
): T {
  let x = embedding(tokens, getTensor(weights, "embedding.weight"));
  rec?.record("embed.out", x);

  for (let i = 0; i < dims.n_layer; i++) {
    const n1 = layernorm(x, getTensor(weights, `ln1.${i}.weight`), getTensor(weights, `ln1.${i}.bias`));
    rec?.record(`layer${i}.ln1.out`, n1);
    x = add(x, timeMixBlock(weights, i, n1, rec));

    const n2 = layernorm(x, getTensor(weights, `ln2.${i}.weight`), getTensor(weights, `ln2.${i}.bias`));
    rec?.record(`layer${i}.ln2.out`, n2);
    x = add(x, channelMixBlock(weights, i, n2, rec));
  }

  x = layernorm(x, getTensor(weights, "ln_out.weight"), getTensor(weights, "ln_out.bias"));
  rec?.record("final_norm.out", x);

  const logits = linear(x, getTensor(weights, "head.weight"));
  rec?.record("head.logits", logits);
  return logits;
}
